Integrate the units system proposed in issue #144 into Euclid types.

nical opened the pull request:
Here is the implementation of the units system from issue #144.

Other than changing how units are expressed, I took the liberty to change the convention ```Typed<Unit, ScalarType>``` into ```Typed<SalarType, Unit>```. This is to be able to use default generic parameters for the units. Default generics in rust are not quite where they should be yet, so the library does not take full advantage of it. When the language issues get resolved, there are some simplifications that can be made to the library (for instance not need both a TypedPoint2D and a Point2D alias).

Member access (```foo.x```) provides scalars (like f32) by default, and there are a few ```_typed``` convenience methods to access members as a Length (```foo.x_typed()```). converting rust-layers to this system, I haven't come across places where the members were accessed without unwrapping the Length right away, which tends to confirm that scalar member access should be what is convenient to write by default.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/euclid/148)
<!-- Reviewable:end -->

nical commented:
r? @metajack

nical commented:
@SimonSapin told me to r? @pcwalton instead.

pcwalton commented on src/matrix2d.rs:
```diff
@@ -106,7 +118,28 @@ impl<T:Add<T, Output=T> +
+}
+
+// Convenient aliases for TypedPoint2D with typed units
+impl<T:Copy, Src, Dst> TypedMatrix2D<T, Src, Dst> {
```
nit: put a space after `:`

pcwalton commented on src/matrix4d.rs:
```diff
@@ -9,69 +9,84 @@
+       PartialOrd +
+       Sub<T, Output=T> +
+       Trig +
+       Zero, Src, Dst> TypedMatrix4D<T, Src, Dst> {
```
Please use a where clause here :)

pcwalton commented on src/matrix4d.rs:
```diff
@@ -355,32 +388,56 @@ impl <T:Add<T, Output=T> +
+    }
+}
 
+impl<T: Copy+fmt::Debug, Src, Dst> fmt::Debug for TypedMatrix4D<T, Src, Dst> {
```
uber-nit: Put spaces around `+`

pcwalton commented on src/point.rs:
```diff
@@ -7,347 +7,497 @@
 
+// Convenient aliases for TypedPoint2D with typed units
+
+impl<T:Clone, U> TypedPoint2D<T, U> {
```
nit: space after `:`

pcwalton commented on src/rect.rs:
```diff
@@ -102,103 +118,128 @@ impl<T: Copy + Clone + PartialOrd + Add<T, Output=T> + Sub<T, Output=T>> Rect<T>
 }
 
-impl<T: Copy + Clone + PartialOrd + Add<T, Output=T> + Sub<T, Output=T> + Zero> Rect<T> {
+impl<T: Copy + Clone + PartialOrd + Add<T, Output=T> + Sub<T, Output=T> + Zero, U> TypedRect<T, U> {
```
While you're here, could you change this to use a `where` clause?

pcwalton commented on src/side_offsets.rs:
```diff
@@ -10,41 +10,54 @@
-#[cfg(feature = "unstable")]
-use heapsize::HeapSizeOf;
 use num::Zero;
+use length::{ Length, Untyped };
```
nit: no spaces after `{` and before `}`

pcwalton commented on src/size.rs:
```diff
@@ -7,117 +7,158 @@
 
-use length::Length;
+use length::{ Length, Untyped };
+use scale_factor::{ ScaleFactor };
```
nit: no spaces after `{` and before `}`, and also no need to brace `{ScaleFactor}`

nical commented:
Thanks for the quick review! I addressed all of the review items and also changed "Untyped" into "UnknownUnit" Sorry I did that so late in the review process. I just realized gecko had renamed its default unit this way a while ago and the name is more understandable I think.

pcwalton commented:
Looks good! Squash the commits together if you'd like and I'll r=me.

To anyone interested: If there are any comments about this change, speak now or forever hold your peace :)

nical commented:
Commits squashed!

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 6205104 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 62051044a723bcd1d3817169734960ca28e6500b -->

bors-servo commented:
:hourglass: Testing commit 6205104 with merge 547b5c3...

bors-servo commented:
:sunny: Test successful - [travis](https://travis-ci.org/servo/euclid/builds/148357870)
